control/controller.go

```diff
@@ -204,32 +243,51 @@ func (c *Controller) enqueueQuery(q *Query) error {
 		return errors.New("failed to transition query to queueing state")
 	}
 
-	// TODO(jsternberg): We should have a real queue! It should only
-	// start the query when we know we have a minimum amount of memory
-	// available and at least one goroutine we can use. Since we don't
-	// implement a queue at the moment, the rest just fakes it by
-	// immediately switching to the execute state and then calling
-	// start which should start the query in a new goroutine and return
-	// the underlying results.
-	return c.executeQuery(q)
+	select {
+	case c.queryQueue <- q:
+	default:
+		return errors.New("queue length exceeded")
+	}
+
+	return nil
+}
+
+func (c *Controller) processQueryQueue() {
+	for {
+		select {
+		case <-c.done:
+			return
+		case q := <-c.queryQueue:
+			c.executeQuery(q)
+		}
+	}
 }
 
-func (c *Controller) executeQuery(q *Query) error {
+func (c *Controller) executeQuery(q *Query) {
 	ctx, ok := q.tryExec()
 	if !ok {
-		return errors.New("failed to transition query into executing state")
+		if q.State() == Canceled {
```

Comment:
If the query is canceled because the controller is shutting down, then we can hit this.  I added this to at least partially return an error in that case.  I will file an issue to have better testing for this.

I'm not sure that you're asking for here since the process queue method doesn't return an error since it's running in a goroutine, the only option is to `SetErr` on the query and/or to log the error.

Reply:
Yes. The intention is that `setErr` is called. There's code in `setErr` to handle the part that you mention.

```go
func (q *Query) setErr(err error) {
    q.stateMu.Lock()
    defer q.stateMu.Unlock()

    // We may have this get called when the query is canceled.
    // If that is the case, transition to the canceled state
    // instead and record the error from that since the error
    // we received is probably wrong.
    select {
    case <-q.parentCtx.Done():
        q.transitionTo(Canceled)
        err = q.parentCtx.Err()
    default:
        q.transitionTo(Errored)
    }
    q.err = err

    // Close the ready channel to report that no results
    // will be sent.
    close(q.results)
}
```

This code also ensures that the results are closed which is required and was missing from your code. I would have forgotten about it if I hadn't looked at the `setErr` method which reminded me that was needed.

Reply:
If you want to add a different error message for the controller being shutdown, this is also probably the correct place to put it within the `select`.

Reply:
Ok, I'll change it to just call `SetErr`.  It would be good to have unit tests that verify what happens when a query is canceled via cient vs. shutdown, but writing those tests is a little tricky, and I want to merge this.